### ISSUE TITLE
Apollo 2.3.0 does not trigger cache changes when cache is populated first time.

### ISSUE THREAD

**Issue by piotr-pawlowski** [apollographql/apollo-android#2531]
**Summary**
Apollo 2.3.0 does not trigger cache changes when cache is populated first time. On 2.2.0 version everything is fine.

**Version**
2.3.0.

**Comment by martinbonnin** [comment]
Thanks for the bug report! Can you share more details about the scenario? 

- How is the cache populated? Through a query? Or using the imperative store APIs (`store.write`, ...) ?
- How are changes detected? A query watcher? Something else?

**Comment by piotr-pawlowski** [comment]
The cache is populated from query and changes are detected with ApolloStore.RecordChangeSubscriber.

**Comment by martinbonnin** [comment]
This might be linked to https://github.com/apollographql/apollo-android/issues/2537 although that other issues apparently also happens with 2.2.0

**Comment by piotr-pawlowski** [comment]
I have found the problem:

In RealApolloStore:

```
@Override public void publish(@NotNull final Set<String> changedKeys) {
    checkNotNull(changedKeys, "changedKeys == null");

    if (changedKeys.isEmpty()) {
      return;
    }

    Set<RecordChangeSubscriber> iterableSubscribers;
    synchronized (this) {
      iterableSubscribers = new LinkedHashSet<>(subscribers);
    }

    for (RecordChangeSubscriber subscriber : iterableSubscribers) {
      subscriber.onCacheRecordsChanged(changedKeys);
    }
  }
```

`changedKeys` is empty.

**Comment by martinbonnin** [comment]
@piotr-pawlowski there is a potential fix in https://github.com/apollographql/apollo-android/pull/2538. 
Can you check the snapshots to see if it applies here?

**Comment by piotr-pawlowski** [comment]
I tested version `2.3.1` and it doesnt work too :(

**Comment by martinbonnin** [comment]
This is expected, the patch isn't released yet. The latest snapshot is `2.3.2-SNAPSHOT`. To use it, you'll have to enable the snapshots repository: https://github.com/apollographql/apollo-android#releases

```
repositories {
  maven { 
    url = uri("https://oss.sonatype.org/content/repositories/snapshots/")
  }
}
```

**Comment by piotr-pawlowski** [comment]
Ok, version `2.3.2-SNAPSHOT` seems to work properly. 
Are you gonna fix the issue #2532 too?

**Comment by martinbonnin** [comment]
> Ok, version 2.3.2-SNAPSHOT seems to work properly.

Nice! Glad to hear that

> Are you gonna fix the issue #2532 too?

This bug looks different, let's move the discussion over there.
